File: android/src/main/java/com/apollohg/editor/tables/EditorCellSelection.kt
package com.apollohg.editor.tables

import com.apollohg.editor.exactV2U32
import org.json.JSONObject

internal sealed interface EditorCellSelection {
    val tableId: String

    data class Drawable(override val tableId: String, val sourcePositions: Set<Int>) : EditorCellSelection
    data class Unavailable(override val tableId: String) : EditorCellSelection
}

internal fun resolveEditorCellSelection(
    selection: JSONObject,
    records: Map<String, JSONObject>
): EditorCellSelection? {
    if (selection.opt("type") != "cell" ||
        selection.keys().asSequence().toSet() != setOf("type", "anchorCell", "headCell")
    ) return null
    val anchor = exactV2U32(selection.opt("anchorCell") as? Number)?.toLong() ?: return null
    val head = exactV2U32(selection.opt("headCell") as? Number)?.toLong() ?: return null

    data class Cell(val position: Int, val row: Int, val column: Int, val rowEnd: Int, val columnEnd: Int) {
        fun intersects(top: Int, left: Int, bottom: Int, right: Int): Boolean =
            row < bottom && rowEnd > top && column < right && columnEnd > left
    }

    val candidates = records.mapNotNull { (id, record) ->
        val tableStart = exactV2U32(record.opt("tablePos") as? Number)?.toLong() ?: return@mapNotNull null
        val tableEnd = exactV2U32(record.opt("sourceEnd") as? Number)?.toLong() ?: return@mapNotNull null
        if (!record.isNull("failure")) {
            return@mapNotNull if (anchor in tableStart until tableEnd && head in tableStart until tableEnd) {
                (tableEnd - tableStart) to EditorCellSelection.Unavailable(id)
            } else null
        }
        val rowCount = exactV2U32(record.opt("rows") as? Number)?.toLong()
            ?.takeIf { it <= Int.MAX_VALUE }?.toInt() ?: return@mapNotNull null
        val columnCount = exactV2U32(record.opt("columns") as? Number)?.toLong()
            ?.takeIf { it <= Int.MAX_VALUE }?.toInt() ?: return@mapNotNull null
        val rawCells = record.optJSONArray("cells") ?: return@mapNotNull null
        val cells = (0 until rawCells.length()).map { index ->
            val cell = rawCells.optJSONObject(index) ?: return@mapNotNull null
            val position = exactV2U32(cell.opt("sourcePos") as? Number)?.toLong()
                ?.takeIf { it <= Int.MAX_VALUE }?.toInt() ?: return@mapNotNull null
            val row = exactV2U32(cell.opt("row") as? Number)?.toLong() ?: return@mapNotNull null
            val column = exactV2U32(cell.opt("column") as? Number)?.toLong() ?: return@mapNotNull null
            val rowspan = exactV2U32(cell.opt("rowspan") as? Number)?.toLong() ?: return@mapNotNull null
            val colspan = exactV2U32(cell.opt("colspan") as? Number)?.toLong() ?: return@mapNotNull null
            if (rowspan == 0L || colspan == 0L || row + rowspan > rowCount.toLong() ||
                column + colspan > columnCount.toLong() || position.toLong() !in tableStart until tableEnd
            ) return@mapNotNull null
            Cell(position, row.toInt(), column.toInt(), (row + rowspan).toInt(), (column + colspan).toInt())
        }
        val first = cells.singleOrNull { it.position.toLong() == anchor } ?: return@mapNotNull null
        val last = cells.singleOrNull { it.position.toLong() == head } ?: return@mapNotNull null
        var top = minOf(first.row, last.row)
        var left = minOf(first.column, last.column)
        var bottom = maxOf(first.rowEnd, last.rowEnd)
        var right = maxOf(first.columnEnd, last.columnEnd)
        while (true) {
            val prior = listOf(top, left, bottom, right)
            cells.filter { it.intersects(top, left, bottom, right) }.forEach { cell ->
                top = minOf(top, cell.row)
                left = minOf(left, cell.column)
                bottom = maxOf(bottom, cell.rowEnd)
                right = maxOf(right, cell.columnEnd)
            }
            if (prior == listOf(top, left, bottom, right)) break
        }
        val selected = cells.filter { it.intersects(top, left, bottom, right) }
            .map { it.position }.toSet()
        (tableEnd - tableStart) to EditorCellSelection.Drawable(id, selected)
    }
    val drawable = candidates.mapNotNull { it.second as? EditorCellSelection.Drawable }
    if (drawable.size > 1) return null
    if (drawable.size == 1) return drawable.single()
    val nearest = candidates.minOfOrNull { it.first } ?: return null
    return candidates.singleOrNull { it.first == nearest }?.second
}

File: ios/Tables/EditorCellSelection.swift
import Foundation

enum EditorCellSelection: Equatable {
    case drawable(tableID: String, sourcePositions: Set<Int>)
    case unavailable(tableID: String)

    private struct Cell {
        let sourcePosition: UInt32
        let top: UInt32
        let left: UInt32
        let bottom: UInt32
        let right: UInt32

        func intersects(_ bounds: Bounds) -> Bool {
            top < bounds.bottom && bottom > bounds.top && left < bounds.right && right > bounds.left
        }
    }

    private struct Bounds: Equatable {
        var top: UInt32
        var left: UInt32
        var bottom: UInt32
        var right: UInt32

        mutating func include(_ cell: Cell) {
            top = min(top, cell.top)
            left = min(left, cell.left)
            bottom = max(bottom, cell.bottom)
            right = max(right, cell.right)
        }
    }

    static func resolve(_ value: Any, records: [String: [String: Any]]) -> EditorCellSelection? {
        guard let selection = value as? [String: Any],
              Set(selection.keys) == ["type", "anchorCell", "headCell"],
              selection["type"] as? String == "cell",
              let anchor = v2ExactUInt32(selection["anchorCell"] as? NSNumber),
              let head = v2ExactUInt32(selection["headCell"] as? NSNumber)
        else { return nil }

        var drawable: [EditorCellSelection] = []
        var unavailable: [(extent: UInt32, selection: EditorCellSelection)] = []
        for (tableID, record) in records {
            guard let tableStart = EditorV2Adapter.uint32Field(record, "tablePos"),
                  let tableEnd = EditorV2Adapter.uint32Field(record, "sourceEnd"),
                  tableStart < tableEnd
            else { continue }
            if !(record["failure"] is NSNull) {
                if tableStart <= anchor && anchor < tableEnd && tableStart <= head && head < tableEnd {
                    unavailable.append((tableEnd - tableStart, .unavailable(tableID: tableID)))
                }
                continue
            }
            guard let rows = EditorV2Adapter.uint32Field(record, "rows"),
                  let columns = EditorV2Adapter.uint32Field(record, "columns"),
                  let rawCells = record["cells"] as? [[String: Any]]
            else { continue }
            var cells: [Cell] = []
            for raw in rawCells {
                guard let source = EditorV2Adapter.uint32Field(raw, "sourcePos"),
                      let row = EditorV2Adapter.uint32Field(raw, "row"),
                      let column = EditorV2Adapter.uint32Field(raw, "column"),
                      let rowspan = EditorV2Adapter.uint32Field(raw, "rowspan"),
                      let colspan = EditorV2Adapter.uint32Field(raw, "colspan"),
                      rowspan > 0, colspan > 0,
                      row < rows, column < columns,
                      rowspan <= rows - row, colspan <= columns - column,
                      tableStart < source, source < tableEnd
                else { return nil }
                cells.append(Cell(sourcePosition: source, top: row, left: column,
                                  bottom: row + rowspan, right: column + colspan))
            }
            guard cells.filter({ $0.sourcePosition == anchor }).count == 1,
                  cells.filter({ $0.sourcePosition == head }).count == 1,
                  let first = cells.first(where: { $0.sourcePosition == anchor }),
                  let last = cells.first(where: { $0.sourcePosition == head })
            else { continue }
            var bounds = Bounds(top: min(first.top, last.top), left: min(first.left, last.left),
                                bottom: max(first.bottom, last.bottom), right: max(first.right, last.right))
            while true {
                let previous = bounds
                for cell in cells where cell.intersects(previous) { bounds.include(cell) }
                if bounds == previous { break }
            }
            drawable.append(.drawable(tableID: tableID, sourcePositions: Set(cells.filter { $0.intersects(bounds) }.map { Int($0.sourcePosition) })))
        }
        if drawable.count == 1 { return drawable[0] }
        if !drawable.isEmpty { return nil }
        guard let smallest = unavailable.map(\.extent).min(),
              unavailable.filter({ $0.extent == smallest }).count == 1
        else { return nil }
        return unavailable.first { $0.extent == smallest }?.selection
    }
}
